src/commands/recover.py

```diff
@@ -122,106 +125,116 @@ async def main(
     validators = await _fetch_registered_validators()
     if not validators:
         raise click.ClickException('No registered validators')
-    click.secho(f'Found {len(validators)} validators, recovering...')
+
+    total_validators = len(validators)
+    click.confirm(
+        f'Vault has {total_validators} registered validator(s), '
+        f'recover active keystores from provided mnemonic?',
+        default=True,
+        abort=True,
+    )
+
+    if keystores_dir.exists():
+        click.confirm(
+            f'Clean up existing {keystores_dir} keystores directory?',
+            default=True,
+            abort=True,
+        )
+        for file in keystores_dir.glob('*'):
+            file.unlink()
 
     mnemonic_next_index = await _generate_keystores(
-        mnemonic,
-        keystores_dir,
-        password_file,
-        validators,
-        per_keystore_password,
+        mnemonic=mnemonic,
+        keystores_dir=keystores_dir,
+        password_file=password_file,
+        validator_statuses=validators,
+        per_keystore_password=per_keystore_password,
     )
 
     config.save(settings.network, mnemonic, mnemonic_next_index)
     click.secho(
-        f'Keystores for vault {settings.vault} successfully recovered to {keystores_dir}',
-        bold=True,
-        fg='green',
+        f'Successfully recovered {greenify(mnemonic_next_index)} '
+        f'keystores for vault {greenify(settings.vault)}',
     )
 
 
 # pylint: disable-next=too-many-locals
-async def _fetch_registered_validators() -> dict[str, str]:
+async def _fetch_registered_validators() -> dict[HexStr, ValidatorStatus | None]:
     """Fetch registered validators."""
-    block = await execution_client.eth.get_block('latest')
-    current_block = block['number']
-    keeper_genesis_block = settings.network_config.KEEPER_GENESIS_BLOCK
-
-    page_size = 50_000
-    public_keys = []
-
-    for cursor in range(keeper_genesis_block, current_block, page_size):
-        page_start = cursor
-        page_end = min(cursor + page_size - 1, current_block)
-
-        events = await vault_contract.events.ValidatorRegistered.get_logs(
-            fromBlock=page_start, toBlock=page_end
-        )
-        for event in events:
-            hex_key = event['args']['publicKey'].hex()
-            public_keys.append(add_0x_prefix(hex_key))
+    click.secho('Fetching registered validators...', bold=True)
+    public_keys = await vault_contract.get_registered_validators_public_keys()
 
-    validators_status = {}
+    if vault_contract.contract_address == settings.network_config.GENESIS_VAULT_CONTRACT_ADDRESS:
+        public_keys.extend(await v2_pool_contract.get_registered_validators_public_keys())
+    click.secho(f'Fetched {len(public_keys)} registered validators', bold=True)
 
+    click.secho('Fetching validators statuses...', bold=True)
+    validator_statuses: dict[HexStr, ValidatorStatus | None] = {
+        public_key: None for public_key in public_keys
+    }
```

Comment:
why dont you start with 
```validator_statuses: dict[HexStr, ValidatorStatus] = {} ``` 
?

Reply:
Because we need public keys that are not yet in the beacon chain (they appear there after 1024 blocks)